### DISCUSSION TITLE
Can you run quarto render with local copies of the CSS and JS?

### DISCUSSION THREAD

**Discussion by fountaindive** [quarto-dev/quarto-cli#11528] [Q&A] [answered]
### Description

For my use case, the machine I view the results of `quarto render` on doesn't have internet access and therefore I can't use any of the web-based methods to get the CSS and JS code.

Is there a way to provide these libraries locally?

Many thanks!

**Comment by cderv** [comment]
> Is there a way to provide these libraries locally?

Which libraries exactly ? 

`quarto render` may require internat access if you try to use `embed-resources` so that those resources are downloaded. 

Otherwise, it will use some CSS and JS libraries usually from CDN based on what is in the template, or based on which R or Python libraries you are using that could provide also some JS dependencies. 

Usually this require internet access are viewing time to load everything in browser from the HTML rendered by Quarto. 

Did you encountered a specific issue ?

**Reply by fountaindive** [reply]
Thanks for the reply!

I'm quite inexperienced when it comes to web development stuff so might be using the wrong terminology sorry.

From what gathered I would require local versions of the files in `site_libs`?

Things like bootstrap, random js that you have to do nice things like navigation and search bar, in the future I'd like to try and have interactive plots using plotly.

Hopefully that makes sense!

**Reply by cderv** [reply, marked as answer]
> From what gathered I would require local versions of the files in site_libs?

Part of what Quarto uses will be shipped with Quarto. This means that doing `quarto render` will copy the right files into `site_libs` for a quarto website for example; This means bootstrap won't be downloaded for example. 

> in the future I'd like to try and have interactive plots using plotly.

For plotly, it won't be shipped with Quarto. it will be provided by other tooling like Python or R libaries that will either bundle or linked to online plotly JS. 

I would say that doing `quarto render` on an environment without internet should mostly work for HTML. Though previewing on same environment will require internet access some of the library (like Mathjax or plotly you mentioned) maybe be retrieved from a CDN so online

**Reply by fountaindive** [reply]
Ah OK, thanks for this it's really helpful. Probably the issue is on my side then!

Best wishes.

**Reply by cderv** [reply]
I may have missed something so don't hesitate to share more while you keep trying using Quarto.
